core/chains/evm/logpoller/log_poller.go

```diff
@@ -612,12 +617,79 @@ func (lp *logPoller) LatestLogEventSigsAddrsWithConfs(fromBlock int64, eventSigs
 
 // GetBlocks tries to get the specified block numbers from the log pollers
 // blocks table. Returns the blocks it was able to find, empty slice if none.
-// It is not guaranteed to find the blocks specified, since
-// the log poller does not save blocks when backfilling.
-// It can be used as a cache of recent blocks (falling back to an RPC on cache miss),
-// to speed up querying and avoid loading an EVM node.
-func (lp *logPoller) GetBlocks(numbers []uint64, qopts ...pg.QOpt) ([]LogPollerBlock, error) {
-	return lp.orm.GetBlocks(numbers, qopts...)
+// When the log poller does not have requested blocks, it falls back
+// to RPC to fetch the missing blocks.
+// response contains blocks in the same order as "numbers" in request parameters
+func (lp *logPoller) GetBlocks(ctx context.Context, numbers []uint64, qopts ...pg.QOpt) ([]LogPollerBlock, error) {
+	blocksFound := make(map[uint64]LogPollerBlock)
+
+	lpBlocks, err := lp.orm.GetBlocks(numbers, qopts...)
+	if err != nil {
+		lp.lggr.Warnw("Error while retrieving blocks from log pollers blocks table. Falling back to RPC...", "requestedBlocks", numbers)
+	}
+	for _, b := range lpBlocks {
+		blocksFound[uint64(b.BlockNumber)] = b
+	}
+
+	// Fallback to RPC for blocks not found in log poller blocks table
+	var reqs []rpc.BatchElem
+	for _, num := range numbers {
+		if _, ok := blocksFound[num]; !ok {
+			req := rpc.BatchElem{
+				Method: "eth_getBlockByNumber",
+				Args:   []interface{}{hexutil.EncodeBig(big.NewInt(0).SetUint64(num)), false},
+				Result: &evmtypes.Head{},
```

Comment:
just double checking - responses from various EVM chains all conform to this type?

Reply:
yes I believe it should be supported. `Head` type is used by `core/chains/evm/client` [here](https://github.com/smartcontractkit/chainlink/blob/develop/core/chains/evm/client/client.go#L251)